test/compliance/test_kaldi.py

```diff
@@ -0,0 +1,140 @@
+
+    if snip_edges:
+        assert(start_sample >= sample_offset and end_sample <= num_samples)
+    else:
+        assert(sample_offset == 0 or start_sample >= sample_offset)
+
+    wave_start = start_sample - sample_offset
+    wave_end = wave_start + frame_length
+    if wave_start >= 0 and wave_end <= wave.size(0):
+        window[f, :] = wave[wave_start:(wave_start + frame_length)]
+    else:
+        wave_dim = wave.size(0)
+        for s in range(frame_length):
+            s_in_wave = s + wave_start
+            while s_in_wave < 0 or s_in_wave >= wave_dim:
+                if s_in_wave < 0:
+                    s_in_wave = - s_in_wave - 1
+                else:
+                    s_in_wave = 2 * wave_dim - 1 - s_in_wave
+            window[f, s] = wave[s_in_wave]
+
+
+class Test_Kaldi(unittest.TestCase):
+    test_dirpath, test_dir = test.common_utils.create_temp_assets_dir()
+    test_filepath = os.path.join(test_dirpath, 'assets', 'kaldi_file.wav')
+
+    def _test_get_strided_helper(self, num_samples, window_size, window_shift, snip_edges):
+        waveform = torch.arange(num_samples).float()
+        output = kaldi._get_strided(waveform, window_size, window_shift, snip_edges)
+
+        # from NumFrames in feature-window.cc
+        n = window_size
+        if snip_edges:
+            m = 0 if num_samples < window_size else 1 + (num_samples - window_size) // window_shift
+        else:
+            m = (num_samples + (window_shift // 2)) // window_shift
+
+        self.assertTrue(output.dim() == 2)
+        self.assertTrue(output.shape[0] == m and output.shape[1] == n)
+
+        window = torch.empty((m, window_size))
+
+        for r in range(m):
+            extract_window(window, waveform, r, window_size, window_shift, snip_edges)
+        self.assertTrue(torch.allclose(window, output))
+
+    def test_get_strided(self):
+        # generate any combination where 0 < window_size <= num_samples and
+        # 0 < window_shift.
+        for num_samples in range(1, 20):
+            for window_size in range(1, num_samples + 1):
+                for window_shift in range(1, 2 * num_samples + 1):
+                    for snip_edges in range(0, 2):
+                        self._test_get_strided_helper(num_samples, window_size, window_shift, snip_edges)
+
+    def _create_data_set(self):
+        # used to generate the dataset to test on. this is not used in testing (offline procedure)
+        test_dirpath = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+        test_filepath = os.path.join(test_dirpath, 'assets', 'kaldi_file.wav')
+        sr = 16000
+        x = torch.arange(0, 20).float()
+        # between [-6,6]
+        y = torch.cos(2 * math.pi * x) + 3 * torch.sin(math.pi * x) + 2 * torch.cos(x)
+        # between [-2^30, 2^30]
+        y = (y / 6 * (1 << 30)).long()
+        # clear the last 16 bits because they aren't used anyways
+        y = ((y >> 16) << 16).float()
```

Comment:
Should we write a compliance load and write based on this? Otherwise I'd expect people to be confused why torchaudio.load + kaldi spectogram doesn't give you the same results as kaldi spectogram directly

Reply:
for torchaudio load if you set normalization=(1<<16) you will read exactly what is in the wav file because all the samples are stored as consecutive 16 bits. therefore, loading from a wav file and what kaldi reads is exactly the same.

for torchaudio save however, we must write with precision=16 (kaldi throws an error ie does not support other formats). the input to torchaudio save is not necessarily what will be saved as the 16 bits due to sox. 
https://github.com/rbouqueau/SoX/blob/master/src/wavpack.c#L138
https://github.com/chirlu/sox/blob/4927023d0978615c74e8a511ce981cf4c29031f1/src/sox.h#L800
A summary of the mapping would be:
```
// x > 0b1111111111111110111111111111111 or 2147450879
if(x > (1 << 31) - 1 - (1 << 15)) {
	// '0b111111111111111' or 32767
	return (1 << 15) - 1
} else {
	return (x + (1 << 15)) >> 16
}
```
so it takes the first 16 bits and adds one if the 17th bit is one and if the summation doesn't cause overflow.

I don't think we need a compliance load because the normalization is pretty simple/straight forward. For compliance write, it is a question of what the user expects those 16 bits to be. They could assume 32 bit to 16 bit sox mapping or they could make sure their input is just 16 bits and within the first 16 bits of 32 so that their input matches exactly what is written to file. Both cases are not kaldi specific though.

Reply:
Even if the wrapper is simple / small, it's still good to add even if it's effectively just documentation.

For saving, we could throw a warning if the given format doesn't match a convention we'll simply choose and fix for consistency.

Reply:
Done.